README.pod
=pod

=encoding UTF-8

=for :stopwords Apocalypse cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee
diff irc mailto metadata placeholders metacpan ARG DBI Kwalitee OID
PostgreSQL SQL SimpleDBI's com github

=for Pod::Coverage *EVERYTHING*

=head1 NAME

POE::Component::SimpleDBI - Asynchronous non-blocking DBI calls in POE made simple

=head1 VERSION

  This document describes v1.31 of POE::Component::SimpleDBI - released November 05, 2014 as part of POE-Component-SimpleDBI.

=head1 SYNOPSIS

	use POE;
	use POE::Component::SimpleDBI;

	# Create a new session with the alias we want
	POE::Component::SimpleDBI->new( 'SimpleDBI' ) or die 'Unable to create the DBI session';

	# Create our own session to communicate with SimpleDBI
	POE::Session->create(
		inline_states => {
			_start => sub {
				# Tell SimpleDBI to connect
				$_[KERNEL]->post( 'SimpleDBI', 'CONNECT',
					'DSN'		=>	'DBI:mysql:database=foobaz;host=192.168.1.100;port=3306',
					'USERNAME'	=>	'FooBar',
					'PASSWORD'	=>	'SecretPassword',
					'EVENT'		=>	'conn_handler',
				);

				# Execute a query and return number of rows affected
				$_[KERNEL]->post( 'SimpleDBI', 'DO',
					'SQL'		=>	'DELETE FROM FooTable WHERE ID = ?',
					'PLACEHOLDERS'	=>	[ qw( 38 ) ],
					'EVENT'		=>	'deleted_handler',
					'INSERT_ID'	=>	0,
				);

				# Retrieve one row of information
				$_[KERNEL]->post( 'SimpleDBI', 'SINGLE',
					'SQL'		=>	'Select * from FooTable LIMIT 1',
					'EVENT'		=>	'success_handler',
					'BAGGAGE'	=>	'Some Stuff I want to keep!',
				);

				# We want many rows of information + get the query ID so we can delete it later
				# Furthermore, disable prepare_cached on this query
				my $id = $_[KERNEL]->call( 'SimpleDBI', 'MULTIPLE',
					'SQL'		=>	'SELECT foo, baz FROM FooTable2 WHERE id = ?',
					'PLACEHOLDERS'	=>	[ qw( 53 ) ],
					'EVENT'		=>	'multiple_handler',
					'PREPARE_CACHED'=>	0,
				);

				# Quote something and send it to another session
				$_[KERNEL]->post( 'SimpleDBI', 'QUOTE',
					'SQL'		=>	'foo$*@%%sdkf"""',
					'SESSION'	=>	'OtherSession',
					'EVENT'		=>	'quote_handler',
				);

				# Changed our mind!
				$_[KERNEL]->post( 'SimpleDBI', 'Delete_Query', $id );

				# 3 ways to shutdown

				# This will let the existing queries finish, then shutdown
				$_[KERNEL]->post( 'SimpleDBI', 'shutdown' );

				# This will terminate when the event traverses
				# POE's queue and arrives at SimpleDBI
				$_[KERNEL]->post( 'SimpleDBI', 'shutdown', 'NOW' );

				# Even QUICKER shutdown :)
				$_[KERNEL]->call( 'SimpleDBI', 'shutdown', 'NOW' );
			},

			# Define your request handlers here
			'quote_handler'	=>	\&FooHandler,
			# And so on
		},
	);

=head1 DESCRIPTION

This module simplifies DBI usage in POE's multitasking world.

This module is a breeze to use, you'll have DBI calls in your POE program
up and running in only a few seconds of setup.

This module does what XML::Simple does for the XML world.

This module works its magic by creating a new session with POE, then spawning off a child process
to do the "heavy" lifting. That way, your main POE process can continue servicing other clients.
Queries are put into a queue, and processed one at a time.

The standard way to use this module is to do this:

	use POE;
	use POE::Component::SimpleDBI;

	POE::Component::SimpleDBI->new( ... );

	POE::Session->create( ... );

	POE::Kernel->run();

=head2 Starting SimpleDBI

To start SimpleDBI, just call it's new method:

	POE::Component::SimpleDBI->new( 'ALIAS' );

This method will die on error or return success.

NOTE: The act of starting/stopping SimpleDBI fires off _child events, read
the POE documentation on what to do with them :)

This constructor accepts only 3 arguments.

=head3 Alias

This sets the session alias in POE.

The default is "SimpleDBI".

=head3 PREPARE_CACHED

This sets the global PREPARE_CACHED setting. This is a boolean value.

	POE::Component::SimpleDBI->new( 'ALIAS', 0 );

The default is enabled.

=head3 SYNCHRONOUS_MODE

This disables the fork() that the subprocess does. Use this only if you are having issues with the backend
and want to debug the database without dealing with multiprocess issues.

	POE::Component::SimpleDBI->new( 'ALIAS', 1, 1 );

The default is disabled.

=head2 Commands

There are a few commands you can trigger in SimpleDBI. They are triggered via $_[KERNEL]->post( ... );

=head3 ID

All of the commands except for Delete_Query and shutdown return an id. To get them, do this:
	my $id = $_[KERNEL]->call( 'SimpleDBI', ... );

Afterwards, the id can be used to delete queries, look at Delete_Query for more information.

=head3 Argument errors

All of the commands validate their arguments, and if an error happens ( missing argument, etc ), they will do either:
	- return undef and forget that your request even existed
	- post to the SESSION/EVENT with ERROR present in the data
		NOTE: The data will not have an ID key present

=head3 Explanation of DO/SINGLE/MULTIPLE/QUOTE arguments

They are passed in via the $_[KERNEL]->post( ... );

NOTE: Capitalization is very important!

=over 4

=item C<SQL>

This is the actual SQL line you want SimpleDBI to execute.
You can put in placeholders, this module supports them.

=item C<PLACEHOLDERS>

This is an array of placeholders.

You can skip this if your query does not utilize it.

=item C<SESSION>

This is the session that will get the result

You can skip this, it defaults to the sending session

=item C<EVENT>

This is the event, triggered whenever a query finished.

It will get a hash in ARG0, consult the specific queries on what you will get.

NOTE: If the key 'ERROR' exists in the hash, then it will contain the error string.

=item C<BAGGAGE>

This is a special argument, you can "attach" any kind of baggage to a query.
The baggage will be kept by SimpleDBI and returned to the Event handler intact.

This is good for storing data associated with a query like a client object, etc.

You can skip this if your query does not utilize it.

=item C<PREPARE_CACHED>

This was added recently, to override SimpleDBI's default behavior of using the
$dbh->prepare_cached() function. Setting this to false will use $dbh->prepare() instead.

Some users reported problems with PostgreSQL. After investigation, this turned out to be
some bizarre OID caching issues when the table was updated while the connection is alive.
The quick work-around is to reconnect to the database, but this was not a "sane" solution.

This is a simple boolean value, and if this argument does not exist, SimpleDBI will
use the global setting when calling new().

=item C<INSERT_ID>

This was added recently, to override SimpleDBI's default behavior of using the
$dbh->last_insert_id() function. Setting this to false will disable retrieval of this value.

This is a simple boolean value, and if this argument does not exist, SimpleDBI will default to true.

=back

=head3 C<CONNECT>

This tells SimpleDBI to connect to the database. NOTE: if we are already connected, it will be a
success ( SimpleDBI will not disconnect then connect automatically ). Accepted arguments:

	DSN		->	The DBI DSN string, consult the DBI docs on what this is
	USERNAME	->	The username for the connection
	PASSWORD	->	The password for the connection
	SESSION		->	The session to send the results
	EVENT		->	The event to send the results
	NOW		->	Tells SimpleDBI to bypass the queue and connect NOW!
	CLEAR		->	Tells SimpleDBI to clear the queue and connect NOW!
	AUTO_COMMIT	->	The boolean value we will pass to DBI->connect ( defaults to true )
	CACHEDKIDS	->	Controls the method to cache prepare_cached queries, an arrayref ( defaults to undef )
	BAGGAGE		->	Any extra data to keep associated with this query ( SimpleDBI will not touch it )

NOTE: if the DSN/USERNAME/PASSWORD/SESSION/EVENT does not exist, SimpleDBI assumes you wanted to use
the old connection and will use the cached values ( if you told it to DISCONNECT ). Here's an example on how to
trigger this event:

	$_[KERNEL]->post( 'SimpleDBI', 'CONNECT',
		'DSN'		=>	'DBI:mysql:database=foobaz;host=192.168.1.100;port=3306',
		'USERNAME'	=>	'MyUser',
		'PASSWORD'	=>	'MyPassword',
		'EVENT'		=>	'conn_handler',
		'NOW'		=>	1,
	);

The NOW/CLEAR arguments are special, they will tell SimpleDBI to bypass the request queue and connect NOW...
The CLEAR argument will also delete all the requests waiting in the queue, they will get an ERROR result.
They both default to false, supply a boolean value to turn them on. The Event handler will get a hash in ARG0:

	{
		'ERROR'		=>	exists only if an error occured
		'GONE'		=>	exists only if the server was disconnected and the reconnect failed
		'ACTION'	=>	'CONNECT'
		'ID'		=>	ID of the Query
		'EVENT'		=>	The event the query will respond to
		'SESSION'	=>	The session the query will respond to
	}

	# NOTE: You can do nifty things like this. They all will be executed in the right order!
	$_[KERNEL]->post( 'SimpleDBI', 'CONNECT', 'DSN' => 'DBI:mysql:...', ... );
	$_[KERNEL]->post( 'SimpleDBI', 'DO', ... );
	$_[KERNEL]->post( 'SimpleDBI', 'SINGLE', ... );
	$_[KERNEL]->post( 'SimpleDBI', 'DISCONNECT' );
	$_[KERNEL]->post( 'SimpleDBI', 'CONNECT', 'DSN' => 'DBI:oracle:...', ... );
	$_[KERNEL]->post( 'SimpleDBI', 'MULTIPLE', ... );
	$_[KERNEL]->post( 'SimpleDBI', 'shutdown' );

As of 1.11 SimpleDBI now detects whether the backend lost the connection to the database server. The backend will
automatically reconnect if it happens, but if that fails, an error will be sent to the session/event specified here
with an extra key: 'GONE'. In this state SimpleDBI is deadlocked, any new queries will not be processed until a
CONNECT NOW event is issued! Keep in mind the SINGLE/etc queries WILL NOT receive an error if this happens, the error
goes straight to the CONNECT handler to keep it simple!

As of 1.29 SimpleDBI added better control of the prepare_cached cache. Some users reported that the subprocess' memory
usage was leaking, and in extreme cases reached several gigs! Upon investigation, it was not SimpleDBI's fault but the
way DBI works. What DBI does is cache the statement handle from $dbh->prepare_cached in the $dbh handle. The problem is
that it stays around forever in the default implementation! Perusing the DBI docs revealed that it was possible to tie
this cache to a custom cache module. So I've added the CACHEDKIDS argument, and setting it to an arrayref will enable the

Changes
Revision history for {{$dist->name}}

{{$NEXT}}

1.31 2014-11-05T22:15:06Z UTC
 - Add note about UTF8 connection attributes - thanks RT#43063
 - Cleanup some code in preparation for future changes - thanks RT#48401
 - Updates to the auto-generated files

 [STATISTICS]
 - code churn: 4 files changed, 237 insertions(+), 235 deletions(-)

1.30 2011-02-10T02:58:17Z UTC
 - Converted to Dist::Zilla for the release process!
 - Added SYNCHRONOUS_MODE to new() for easier DB debugging - thanks
   RCAPUTO!

1.29 0000-00-00T00:00:00Z UTC
 - Added better caching control in the SubProcess - resolves a memory leak,
   thanks Andrew Feren and wirelessdreamer!
 - Added the CACHEDKIDS argument to CONNECT

1.28 0000-00-00T00:00:00Z UTC
 - Minor POD tweaks
 - Fixed RT#45948 ( Database doesn't reconnect when child died ) - thanks
   Paul Williams!
 - Fixed RT#48237 ( use of exit causes crashes in threaded code (MsWin32) )
   - thanks Andrew Feren!
 - Fixed RT#48400 ( BAGGAGE not a valid argument for CONNECT ) - thanks
   Andrew Feren!
 - BAGGAGE is now a valid argument for CONNECT/DISCONNECT

1.27 0000-00-00T00:00:00Z UTC
 - Fixed RT#43408 ( fix "Modification of a read-only value attempted at"
   error at shutdown ) - thanks acferen!
 - Fixed RT#43442 ( tweak binmode for MSWin32 ) - thanks acferen!
 - Added link to the github.com repository in the POD
 - Added link to CPAN::Forum in the POD

1.26 0000-00-00T00:00:00Z UTC
 - Kwalitee-related fixes ( moved to Test::Apocalypse )
 - Fixed RT#42890 ( tweaked Wheel::Run arguments on Win32, please yell at
   me if it doesn't work for you! )
 - Added sigCHLD management

1.25 0000-00-00T00:00:00Z UTC
 - Kwalitee-related fixes

1.24 0000-00-00T00:00:00Z UTC
 - Bumped Test::More prereq to 0.86 because I was getting FAIL reports from
   outdated people :(

1.23 0000-00-00T00:00:00Z UTC
 - Switched to Build.PL for the build system
 - Fixed the stupid test dependencies, thanks BiNGOs!
 - Added the new EXPERIMENTAL 'ATOMIC' support, please let me know if it's
   broken on your setup!
 - Added some more author tests
 - Added AUTO_COMMIT argument to CONNECT to control the DBI variable (
   defaults to true )

1.22 0000-00-00T00:00:00Z UTC
 - Kwalitee-related fixes

1.21 0000-00-00T00:00:00Z UTC
 - silence warnings when used with DBD::SQLite - thanks to Sjors Gielen for
   tracking this down!

1.20 0000-00-00T00:00:00Z UTC
 - Added the INSERT_ID to control $dbh->last_insert_id usage

1.19 0000-00-00T00:00:00Z UTC
 - Added the PREPARE_CACHED argument to control caching

1.18 0000-00-00T00:00:00Z UTC
 - Ignore the DBI error for last_insert_id and default to undef

1.17 0000-00-00T00:00:00Z UTC
 - Added "INSERTID" to the result of DO

1.16 0000-00-00T00:00:00Z UTC
 - Noticed a glaring documentation bug - SINGLE queries return mixedCaps
   rows ( not lowercase! ) - MULTIPLE queries return lowercase rows
 - This makes me *VERY* tempted to fix SINGLE to return lowercase, is this
   a good idea? Let me know!
 - Fixed SimpleDBI failure on Win32 - thanks RT #23851

1.15 0000-00-00T00:00:00Z UTC
 - Thanks to Fred Castellano, who stumbled on a DEADLOCK bug, fixed!
 - Added sanity tests to CONNECT/DISCONNECT

1.14 0000-00-00T00:00:00Z UTC
 - learned about the difference between ref $self and ref( $self )
 - Kwalitee-related fixes

1.13 0000-00-00T00:00:00Z UTC
 - Finally use a Changes file - thanks RT #18981
 - Fixed a bug in SINGLE if returned_rows = 0 it will not return undef, but
   give us blank rows!
 - Documentation tweaks

1.12 0000-00-00T00:00:00Z UTC
 - In the SubProcess, added a binmode() to STDIN and STDERR, for the
   windows attempt
 - Added code to make SimpleDBI work in Win32 boxes, thanks to the recent
   Wheel::Run patches!
 - Documentation tweaks as usual

1.11 0000-00-00T00:00:00Z UTC
 - Hannes had a problem: His IRC bot logs events to a database, and
   sometimes there is no events to log after hours and hours of inactivity
   ( must be a boring channel haha ), the db server disconnected!
 - The solution was to do a $dbh->ping() before each query, if your DBI
   driver does it inefficiently, go yell at them!
 - In the event that a reconnect is not possible, an error will be sent to
   the CONNECT event handler, look at the updated pod.

1.10 0000-00-00T00:00:00Z UTC
 - Fixed a bug in the DO routine, thanks to Hannes!

1.09 0000-00-00T00:00:00Z UTC
 - Removed the abstract LIMIT 1 to the SINGLE query
 - Removed the silly 5.8.x requirement in Makefile.PL
 - Made the SubProcess use less memory by exec()ing itself
 - Added the new CONNECT/DISCONNECT commands
 - Removed the db connection information from new()
 - Minor tweaks here and there to not stupidly call() the queue checker
   when there is nothing to check :)
 - Added the sysreaderr debugging output
 - More intelligent SQL/PLACEHOLDERS/BAGGAGE handling
 - Made the command arguments more stricter, it will only accept valid
   arguments, instead of just extracting what it needs
 - Made sure all return data have ID/EVENT/SESSION/ACTION in them for easy
   debugging
 - Added the SESSION parameter to all commands for easy redirection
 - Updated the POD and generally made it better :)
 - Added a new command -> Clear_Queue ( clears the queue )


lib/POE/Component/SimpleDBI/SubProcess.pm
package POE::Component::SimpleDBI::SubProcess;

# ABSTRACT: Backend of POE::Component::SimpleDBI

# Use Error.pm's try/catch semantics
use Error 0.15 qw( :try );

# We pass in data to POE::Filter::Reference
use POE::Filter::Reference;

# We run the actual DB connection here
use DBI 1.30;

# Our Filter object
my $filter = POE::Filter::Reference->new();

# Our DBI handle
my $DB;

# Save the connect struct for future use
my $CONN;

# Sysread error hits
my $sysreaderr = 0;

# Shut up Perl::Critic!
## no critic ( ProhibitAccessOfPrivateData )

# This is the subroutine that will get executed upon the fork() call by our parent
sub main {
	# Autoflush to avoid weirdness
	$|++;

	# set binmode, thanks RT #43442
	binmode( STDIN );
	binmode( STDOUT );

	# Okay, now we listen for commands from our parent :)
	while ( sysread( STDIN, my $buffer = '', 1024 ) ) {
		# Feed the line into the filter
		my $data = $filter->get( [ $buffer ] );

		# INPUT STRUCTURE IS:
		# $d->{'ACTION'}	= SCALAR	->	WHAT WE SHOULD DO
		# $d->{'SQL'}		= SCALAR	->	THE ACTUAL SQL
		# $d->{'SQL'}		= ARRAY		->	THE ACTUAL SQL ( in case of ATOMIC )
		# $d->{'PLACEHOLDERS'}	= ARRAY		->	PLACEHOLDERS WE WILL USE
		# $d->{'PREPARE_CACHED'}= BOOLEAN	->	USE CACHED QUERIES?
		# $d->{'ID'}		= SCALAR	->	THE QUERY ID ( FOR PARENT TO KEEP TRACK OF WHAT IS WHAT )

		# $d->{'DSN'}		= SCALAR	->	DSN for CONNECT
		# $d->{'USERNAME'}	= SCALAR	->	USERNAME for CONNECT
		# $d->{'PASSWORD'}	= SCALAR	->	PASSWORD for CONNECT

		# Process each data structure
		foreach my $input ( @$data ) {
			output( process_request( $input ) );
			return if $input->{'ACTION'} eq 'EXIT';
		}
	}

	# Arrived here due to error in sysread/etc
	output( Make_Error( 'SYSREAD', $! ) );

	# If we got more than 5 sysread errors, abort!
	if ( ++$sysreaderr == 5 ) {
		if ( defined $DB ) { $DB->disconnect() }
		return;
	} else {
		goto &main;
	}

	return;
}

sub process_request {
	my $input = shift;

	# Now, we do the actual work depending on what kind of query it was
	if ( $input->{'ACTION'} eq 'CONNECT' ) {
		# Connect!
		return (DB_CONNECT($input, 0))[1];
	} elsif ( $input->{'ACTION'} eq 'DISCONNECT' ) {
		# Disconnect!
		return DB_DISCONNECT( $input );
	} elsif ( $input->{'ACTION'} eq 'DO' ) {
		# Fire off the SQL and return success/failure + rows affected
		return DB_DO( $input );
	} elsif ( $input->{'ACTION'} eq 'SINGLE' ) {
		# Return a single result
		return DB_SINGLE( $input );
	} elsif ( $input->{'ACTION'} eq 'MULTIPLE' ) {
		# Get many results, then return them all at the same time
		return DB_MULTIPLE( $input );
	} elsif ( $input->{'ACTION'} eq 'QUOTE' ) {
		return DB_QUOTE( $input );
	} elsif ( $input->{'ACTION'} eq 'ATOMIC' ) {
		return DB_ATOMIC( $input );
	} elsif ( $input->{'ACTION'} eq 'EXIT' ) {
		# Cleanly disconnect from the DB
		if ( defined $DB ) {
			$DB->disconnect();
			undef $DB;
		}

		# EXIT!
		return;
	}

	# Unrecognized action!
	return( Make_Error( $input->{'ID'}, 'Unknown action sent from parent' ) );
}

# Connects to the DB
sub DB_CONNECT {
	# Get the input structure
	my $data = shift;

	# Are we reconnecting?
	my $reconn = shift;

	# Our output structure
	my $output;

	# Are we already connected?
	if ( defined $DB and $DB->ping() ) {
		# Output success
		$output = {
			'ID'	=>	$data->{'ID'},
		};
	} else {
		# Actually make the connection :)
		try {
			$DB = DBI->connect(
				# The DSN we just set up
				$data->{'DSN'},

				# Username
				$data->{'USERNAME'},

				# Password
				$data->{'PASSWORD'},

				# We set some configuration stuff here
				{
					# We do not want users seeing 'spam' on the commandline...
					'PrintError'	=>	0,
					'PrintWarn'	=>	0,

					# Automatically raise errors so we can catch them with try/catch
					'RaiseError'	=>	1,

					# Disable the DBI tracing
					'TraceLevel'	=>	0,

					# AutoCommit our stuff?
					'AutoCommit'	=>	$data->{'AUTO_COMMIT'},
				}
			);

			# Check for undefined-ness
			if ( ! defined $DB ) {
				die "Error Connecting: $DBI::errstr";
			} else {
				# Did we request a custom cache module?
				if ( defined $data->{'CACHEDKIDS'} ) {
					eval "require $data->{'CACHEDKIDS'}->[0]";
					die "Unable to load custom caching module: $@" if $@;

					# code lifted from DBI's POD, thanks!
					my $cache;
					tie %$cache, @{ $data->{'CACHEDKIDS'} };
					$DB->{'CachedKids'} = $cache;
				}

				# Output success
				$output = {
					'ID'	=>	$data->{'ID'},
				};

				# Save this!
				$CONN = $data;
			}
		} catch Error with {
			# Get the error
			my $e = shift;

			# Declare it!
			$output = Make_Error( $data->{'ID'}, $e );
		};
	}

	# All done!
	if ( ! defined $reconn ) {
		return (1, $output);
	} else {
		# Reconnect attempt, was it successful?
		if ( ! exists $output->{'ERROR'} ) {
			return (1, $output);
		}
	}

	return (0, $output);
}

# Disconnects from the DB
sub DB_DISCONNECT {
	# Get the input structure
	my $data = shift;

	# Our output structure
	my $output;

	# Are we already disconnected?
	if ( ! defined $DB ) {
		# Output success
		$output = {
			'ID'	=>	$data->{'ID'},
		};
	} else {
		# Disconnect from the DB
		try {
			$DB->disconnect();
			undef $DB;

			# Output success
			$output = {
				'ID'	=>	$data->{'ID'},
			};
		} catch Error with {
			# Get the error
			my $e = shift;

			# Declare it!
			$output = Make_Error( $data->{'ID'}, $e );
		};
	}

	# All done!
	return $output;
}

# This subroutine does a DB QUOTE
sub DB_QUOTE {
	# Get the input structure
	my $data = shift;

	# The result
	my( $quoted, $output );

	# Check if we are connected
	if ( ! defined $DB or ! $DB->ping() ) {
		# Automatically try to reconnect
		if ( ! ( DB_CONNECT( $CONN, 'RECONNECT' ) )[0] ) {
			return Make_Error( 'GONE', 'Lost connection to the database server.' );
		}
	}

	# Quote it!
	try {
		$quoted = $DB->quote( $data->{'SQL'} );
	} catch Error with {
		# Get the error
		my $e = shift;

		$output = Make_Error( $data->{'ID'}, $e );
	};

	# Check for errors
	if ( ! defined $output ) {
		# Make output include the results
		$output = {};
		$output->{'DATA'} = $quoted;
		$output->{'ID'} = $data->{'ID'};
	}

	# All done!
	return $output;
}

# This subroutine runs a 'SELECT' style query on the db
sub DB_MULTIPLE {
	# Get the input structure
	my $data = shift;

	# Variables we use
	my( $output, $sth );
	my $result = [];

	# Check if we are connected
	if ( ! defined $DB or ! $DB->ping() ) {
		# Automatically try to reconnect
		if ( ! ( DB_CONNECT( $CONN, 'RECONNECT' ) )[0] ) {
			return( Make_Error( 'GONE', 'Lost connection to the database server.' ) );
		}
	}

	# Catch any errors :)
	try {
		# Make a new statement handler and prepare the query
		if ( $data->{'PREPARE_CACHED'} ) {
			$sth = $DB->prepare_cached( $data->{'SQL'} );
		} else {
			$sth = $DB->prepare( $data->{'SQL'} );
		}

		# Check for undef'ness
		if ( ! defined $sth ) {
			die "Did not get sth: $DBI::errstr";
		} else {
			# Execute the query
			try {
				# Put placeholders?
				if ( exists $data->{'PLACEHOLDERS'} and defined $data->{'PLACEHOLDERS'} ) {
					$sth->execute( @{ $data->{'PLACEHOLDERS'} } );
				} else {
					$sth->execute();
				}
			} catch Error with {
				die $sth->errstr if $sth->errstr;
				die "NON DBI ERROR : $@";
			};
		}

		# The result hash
		my $newdata;

		# Bind the columns
		try {
			$sth->bind_columns( \( @$newdata{ @{ $sth->{'NAME_lc'} } } ) );
		} catch Error with {
			die $sth->errstr if $sth->errstr;
			die "NON DBI ERROR : $@";
		};

		# Actually do the query!
		try {
			while ( $sth->fetch() ) {
				# Copy the data, and push it into the array
				push( @{ $result }, { %{ $newdata } } );
			}
		} catch Error with {
			die $sth->errstr if $sth->errstr;
			die "NON DBI ERROR : $@";
		};

		# Check for any errors that might have terminated the loop early
		if ( $sth->err() ) {
			# Premature termination!
			die $sth->errstr;
		}
	} catch Error with {
		# Get the error
		my $e = shift;

		$output = Make_Error( $data->{'ID'}, $e );
	};

	# Check if we got any errors
	if ( ! defined $output ) {
		# Make output include the results
		$output = {};
		$output->{'DATA'} = $result;
		$output->{'ID'} = $data->{'ID'};
	}

	# Finally, we clean up this statement handle
	if ( defined $sth ) {
		$sth->finish();

		# Make sure the object is gone, thanks Sjors!
		undef $sth;
	}

	# Return the data structure
	return $output;
}

# This subroutine runs a 'SELECT ... LIMIT 1' style query on the db
sub DB_SINGLE {
	# Get the input structure
	my $data = shift;

	# Variables we use
	my( $output, $sth, $result );

	# Check if we are connected
	if ( ! defined $DB or ! $DB->ping() ) {
		# Automatically try to reconnect
		if ( ! ( DB_CONNECT( $CONN, 'RECONNECT' ) )[0] ) {
			return Make_Error( 'GONE', 'Lost connection to the database server.' );
		}
	}

	# Catch any errors :)
	try {
		# Make a new statement handler and prepare the query
		if ( $data->{'PREPARE_CACHED'} ) {
			$sth = $DB->prepare_cached( $data->{'SQL'} );
		} else {
			$sth = $DB->prepare( $data->{'SQL'} );
		}

		# Check for undef'ness
		if ( ! defined $sth ) {
			die "Did not get sth: $DBI::errstr";
		} else {
			# Execute the query
			try {
				# Put placeholders?
				if ( exists $data->{'PLACEHOLDERS'} and defined $data->{'PLACEHOLDERS'} ) {
					$sth->execute( @{ $data->{'PLACEHOLDERS'} } );
				} else {
					$sth->execute();
				}
			} catch Error with {
				die $sth->errstr if $sth->errstr;
				die "NON DBI ERROR : $@";
			};
		}

		# Actually do the query!
		try {
			$result = $sth->fetchrow_hashref();
		} catch Error with {
			die $sth->errstr if $sth->errstr;
			die "NON DBI ERROR : $@";
		};
	} catch Error with {
		# Get the error
		my $e = shift;

		$output = Make_Error( $data->{'ID'}, $e );
	};

	# Check if we got any errors
	if ( ! defined $output ) {
		# Make output include the results
		$output = {};
		$output->{'DATA'} = $result;
		$output->{'ID'} = $data->{'ID'};
	}

	# Finally, we clean up this statement handle
	if ( defined $sth ) {
		$sth->finish();

		# Make sure the object is gone, thanks Sjors!
		undef $sth;
	}

	# Return the data structure
	return $output;
}

# This subroutine runs a 'DO' style query on the db
sub DB_DO {
	# Get the input structure
	my $data = shift;

	# Variables we use
	my( $output, $sth, $rows_affected, $last_id );

	# Check if we are connected
	if ( ! defined $DB or ! $DB->ping() ) {
		# Automatically try to reconnect
		if ( ! ( DB_CONNECT( $CONN, 'RECONNECT' ) )[0] ) {
			return Make_Error( 'GONE', 'Lost connection to the database server.' );
		}
	}

	# Catch any errors :)
	try {
		# Make a new statement handler and prepare the query
		if ( $data->{'PREPARE_CACHED'} ) {
			$sth = $DB->prepare_cached( $data->{'SQL'} );
		} else {
			$sth = $DB->prepare( $data->{'SQL'} );
		}

		# Check for undef'ness
		if ( ! defined $sth ) {
			die "Did not get sth: $DBI::errstr";
		} else {
			# Execute the query
			try {
				# Put placeholders?
				if ( exists $data->{'PLACEHOLDERS'} and defined $data->{'PLACEHOLDERS'} ) {
					$rows_affected = $sth->execute( @{ $data->{'PLACEHOLDERS'} } );
				} else {
					$rows_affected = $sth->execute();
				}

				# Should we even attempt this?
				if ( $data->{'INSERT_ID'} ) {
					try {
						# Get the last insert id ( make this portable! )
						$last_id = $DB->last_insert_id( undef, undef, undef, undef );
					} catch Error with {
						# Ignore this error!
					};
				}
			} catch Error with {
				# If something other than DBI fails, errstr won't be set.
				# You can get this if, for example, we need a module that hasn't been loaded.
				die $sth->errstr if $sth->errstr;
				die "NON DBI ERROR : $@";
			};
		}
	} catch Error with {
		# Get the error
		my $e = shift;

		$output = Make_Error( $data->{'ID'}, $e );
	};

	# If rows_affected is not undef, that means we were successful
	if ( defined $rows_affected && ! defined $output ) {
		# Make the data structure
		$output = {};
		$output->{'DATA'} = $rows_affected;
		$output->{'ID'} = $data->{'ID'};
		$output->{'INSERTID'} = $last_id;
	} elsif ( ! defined $rows_affected && ! defined $output ) {
		# Internal error...
		die 'Internal Error in DB_DO';
	}

	# Finally, we clean up this statement handle
	if ( defined $sth ) {
		$sth->finish();

		# Make sure the object is gone, thanks Sjors!
		undef $sth;
	}

	# Return the data structure
	return $output;
}

# This subroutine runs a 'DO' style query on the db in a transaction
sub DB_ATOMIC {
	# Get the input structure
	my $data = shift;

	# Variables we use
	my( $output, $sth );

	# Check if we are connected
	if ( ! defined $DB or ! $DB->ping() ) {
		# Automatically try to reconnect
		if ( ! ( DB_CONNECT( $CONN, 'RECONNECT' ) )[0] ) {
			return Make_Error( 'GONE', 'Lost connection to the database server.' );
		}
	}

	# Catch any errors :)
	try {
		# start the transaction
		$DB->begin_work if $DB->{'AutoCommit'};

		# process each query
		for my $idx ( 0 .. $#{ $data->{'SQL'} } ) {
			if ( $data->{'PREPARE_CACHED'} ) {
				$sth = $DB->prepare_cached( $data->{'SQL'}->[ $idx ] );
			} else {
				$sth = $DB->prepare( $data->{'SQL'}->[ $idx ] );
			}

			# Check for undef'ness
			if ( ! defined $sth ) {
				die "Did not get sth: $DBI::errstr";
			} else {
				# actually execute it!
				try {
					if ( exists $data->{'PLACEHOLDERS'} and defined $data->{'PLACEHOLDERS'} and defined $data->{'PLACEHOLDERS'}->[ $idx ] ) {
						$sth->execute( @{ $data->{'PLACEHOLDERS'}->[ $idx ] } );
					} else {
						$sth->execute;
					}
				} catch Error with {
					die $sth->errstr;
				};

				# Finally, we clean up this statement handle
				$sth->finish();

				# Make sure the object is gone, thanks Sjors!
				undef $sth;
			}
		}

		# done with transaction!
		$DB->commit;
	} catch Error with {
		# Get the error
		my $e = shift;

		# rollback the changes!
		try {
			$DB->rollback;
		} catch Error with {
			# Get the error
			my $error = shift;

			$output = Make_Error( $data->{'ID'}, 'ROLLBACK_FAILURE: ' . $error . ' on query error: ' . $e );
		};

		# did we rollback fine?
		if ( ! defined $output ) {
			$output = Make_Error( $data->{'ID'}, 'COMMIT_FAILURE: ' . $e );
		}
	};

	# Finally, we clean up this statement handle
	if ( defined $sth ) {
		$sth->finish();

		# Make sure the object is gone, thanks Sjors!
		undef $sth;
	}

	# If we got no output, we did it!
	if ( ! defined $output ) {
		# Make the data structure
		$output = {};
		$output->{'DATA'} = 'SUCCESS';
		$output->{'ID'} = $data->{'ID'};
	}

	# Return the data structure
	return $output;
}

# This subroutine makes a generic error structure
sub Make_Error {
	# Make the structure
	my $data = {};
	$data->{'ID'} = shift;

	# Get the error, and stringify it in case of Error::Simple objects
	my $error = shift;

	if ( ref $error and ref( $error ) eq 'Error::Simple' ) {
		$data->{'ERROR'} = $error->text;
	} else {
		$data->{'ERROR'} = $error;
	}

	# All done!
	return $data;
}

# This subroutine makes a generic DEBUG structure
sub Make_DEBUG {
	# Make the structure
	my $data = {};
	$data->{'ID'} = 'DEBUG';

	# Get the data, and shove it in the hash
	my @debug = @_;
	$data->{'RESULT'} = \@debug;

	# All done!
	return $data;
}

# Prints any output to STDOUT
sub output {
	# Get the data
	my $data = shift;

	# Freeze it!
	my $output = $filter->put( [ $data ] );

	# Print it!
	print STDOUT @$output;
	return;
}

1;

=pod

=for Pod::Coverage *EVERYTHING*

=for stopwords DBI

=head1 DESCRIPTION

This module is responsible for implementing the guts of POE::Component::SimpleDBI.
Namely, the fork/exec and the connection to the DBI.

=cut
